sdk/core/azure-core/src/main/java/com/azure/core/implementation/AsynchronousFileChannelAdapter.java

```diff
@@ -0,0 +1,155 @@
+                }
+                endOperation(Operation.READ);
+                handler.completed(result, attachment);
+            }
+
+            @Override
+            public void failed(Throwable exc, A attachment) {
+                endOperation(Operation.READ);
+                handler.failed(exc, attachment);
+            }
+        });
+    }
+
+    @Override
+    public Future<Integer> read(ByteBuffer dst) {
+        beginOperation(Operation.READ);
+        CompletableFuture<Integer> future = new CompletableFuture<>();
+        fileChannel.read(dst, position.get(), dst, new CompletionHandler<Integer, ByteBuffer>() {
+            @Override
+            public void completed(Integer result, ByteBuffer attachment) {
+                if (result >= 0) {
+                    position.addAndGet(result);
+                }
+                endOperation(Operation.READ);
+                future.complete(result);
+            }
+
+            @Override
+            public void failed(Throwable exc, ByteBuffer attachment) {
+                endOperation(Operation.READ);
+                future.completeExceptionally(exc);
+            }
+        });
+        return future;
+    }
+
+    @Override
+    public <A> void write(ByteBuffer src, A attachment, CompletionHandler<Integer, ? super A> handler) {
+        beginOperation(Operation.WRITE);
+        fileChannel.write(src, position.get(), attachment, new CompletionHandler<Integer, A>() {
+            @Override
+            public void completed(Integer result, A attachment) {
+                position.addAndGet(result);
+                endOperation(Operation.WRITE);
+                handler.completed(result, attachment);
+            }
+
+            @Override
+            public void failed(Throwable exc, A attachment) {
+                endOperation(Operation.WRITE);
+                handler.failed(exc, attachment);
+            }
+        });
+    }
+
+    @Override
+    public Future<Integer> write(ByteBuffer src) {
+        beginOperation(Operation.WRITE);
+        CompletableFuture<Integer> future = new CompletableFuture<>();
+        fileChannel.write(src, position.get(), src, new CompletionHandler<Integer, ByteBuffer>() {
+            @Override
+            public void completed(Integer result, ByteBuffer attachment) {
+                position.addAndGet(result);
+                endOperation(Operation.WRITE);
+                future.complete(result);
+            }
+
+            @Override
+            public void failed(Throwable exc, ByteBuffer attachment) {
+                endOperation(Operation.WRITE);
+                future.completeExceptionally(exc);
+            }
+        });
+        return future;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return fileChannel.isOpen();
+    }
+
+    @Override
+    public void close() throws IOException {
+        fileChannel.close();
+    }
+
+    private void beginOperation(Operation operation) {
```

Comment:
Would it ever be possible for concurrent operations to happen at the same time? For example, in the future we could allow for multiple write operations to be put on to the write queue to allow for them to be "parallelized" (right now we don't allow another write operation until the completion fires)

Reply:
In that case we'd likely allow series of writes and series of reads - the series kinds will still be mutually exclusive but we could allow backlog of operations.

For now this has parity with how do we write to file in fluxutil, so I'd defer this optimization.